aqueduct_client/aqueduct_client.py

```diff
@@ -0,0 +1,243 @@
+            "params": params,
+            "name": name,
+        }
+
+        response = self._post('', args)
+        response.raise_for_status()
+        created_pipeline_id = response.json()['pipeline_id']
+
+        return created_pipeline_id
+
+    def get_results_dataframe(self, pipeline_id):
+        """
+        Gets the result of this pipeline in a pandas dataframe.
+
+        Parameters
+        ----------
+        pipeline_id : str
+            The id of the pipeline whose results should be loaded.
+
+        Returns
+        -------
+        pd.DataFrame
+            A dataframe holding the result, indexed by date and the
+            asset identifier format (symbol, sid, or fsym_region_id)
+            that this pipeline used.
+        """
+        pipeline_status = self.get_pipeline(pipeline_id)
+        if pipeline_status["status"] == "RUNNING":
+            raise ValueError("Pipeline {id} is still running!".format(
+                id=pipeline_id
+            ))
+        elif pipeline_status["status"] == "FAILED":
+            raise ValueError(
+                "Pipeline {id} ended in error, use `get_error` "
+                "to get its error message."
+            ).format(id=pipeline_id)
+
+        asset_identifier_format = pipeline_status["asset_identifier_format"]
+
+        # now that we know the pipeline isn't still running, get its results
+        response = self._get('/{id}/results_url'.format(
+            base=self._base_url,
+            id=pipeline_id
+        ), None)
+
+        url = response.json()['url']
+
+        # get the data from the url
+        results_url_resp = requests.get(url)
+
+        if results_url_resp.status_code != 200:
+            raise ValueError("Could not download results from given url.")
+
+        result_df = pd.read_csv(
+            pd.compat.StringIO(results_url_resp.text),
+            index_col=['date', asset_identifier_format],
+        )
+
+        return result_df
+
+    def get_error(self, id):
+        """
+        Gets the error that caused this pipeline to fail to complete
+        successfully.
+
+        Parameters
+        ----------
+        id : str
+            The id of the pipeline whose errors should be loaded.
+
+        Returns
+        -------
+        dict
+            A dictionary that can contain `date`, `name`, `message`,
+            `lineno`, `method` keys.
+        """
+        pipeline_status = self.get_pipeline(id)
+        if pipeline_status["status"] != "FAILED":
+            raise ValueError(
+                "Pipeline {id} did not end in error!"
+            ).format(id=id)
+
+        # get the error
+        response = self._get(
+            '/{id}/exception'.format(id=id),
+            None
+        )
+
+        response.raise_for_status()
+
+        return response.json()
+
+    def _get(self, path, body, *args, **kwargs):
+        return requests.get(
+            self._base_url + path,
```

Comment:
https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin is probably the slightly safer/cleaner way to do this, but FYI it requires different imports in py2 and py3.

Reply:
seems like a good improvement, will do